File: src/main/java/com/hjj/apiserver/service/PurchaseService.java
package com.hjj.apiserver.service;

import com.hjj.apiserver.common.exception.UserNotFoundException;
import com.hjj.apiserver.domain.*;
import com.hjj.apiserver.dto.CategoryDto;
import com.hjj.apiserver.dto.PurchaseDto;
import com.hjj.apiserver.repositroy.*;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Slice;
import org.springframework.data.domain.SliceImpl;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

@Slf4j
@Service
@Transactional(readOnly = true)
@AllArgsConstructor
public class PurchaseService {

    private final AccountBookRepository accountBookRepository;
    private final CategoryRepository categoryRepository;
    private final PurchaseRepository purchaseRepository;
    private final UserRepository userRepository;
    private final CardRepository cardRepository;
    private final ModelMapper modelMapper;


    @Transactional(readOnly = false, rollbackFor = Exception.class)
    public void addPurchase(PurchaseDto purchaseDto) throws Exception {
        UserEntity userEntity = userRepository.getById(purchaseDto.getUserNo());
        if(userEntity != null){
            purchaseDto.setUserEntity(userEntity);
        }

        AccountBookEntity accountBookEntity = accountBookRepository.findAccountBookBySubQuery(purchaseDto.getUserNo(), purchaseDto.getAccountBookNo()).orElseThrow(UserNotFoundException::new);
        purchaseDto.setAccountBookEntity(accountBookEntity);

        if(purchaseDto.getCardNo() != null){
            CardEntity cardEntity = cardRepository.getById(purchaseDto.getCardNo());
            if(cardEntity == null)
                throw new Exception("해당 되는 카드가 존재하지 않습니다.");
            purchaseDto.setCardEntity(cardEntity);
        }
        if(purchaseDto.getCategoryNo() != null){
            List<AccountBookUserEntity.AccountRole> accountRoleList = new ArrayList<>();
            accountRoleList.add(AccountBookUserEntity.AccountRole.OWNER);
            accountRoleList.add(AccountBookUserEntity.AccountRole.MEMBER);
            CategoryEntity categoryEntity = categoryRepository.findByCategoryNoAndSubQuery(purchaseDto.getCategoryNo(), purchaseDto.getAccountBookNo(), purchaseDto.getUserNo(), accountRoleList).orElseThrow(() -> new Exception("해당하는 카테고리가 없습니다."));
            purchaseDto.setCategoryEntity(categoryEntity);
        }

        PurchaseEntity purchaseEntity = purchaseDto.toEntity();

        purchaseRepository.save(purchaseEntity);

    }

    public List<PurchaseDto.Purchase> findPurchaseList(PurchaseDto purchaseDto){

        List<PurchaseEntity> purchaseEntityList = purchaseRepository.findAllEntityGraphByPurchaseDateBetweenAndAccountBookEntity_AccountBookNoAndDeleteYnOrderByPurchaseDateDesc(purchaseDto.getStartDate(), purchaseDto.getEndDate(), purchaseDto.getAccountBookNo(), 'N');
        List<PurchaseDto.Purchase> purchaseList = new ArrayList<>();
        purchaseEntityList.stream().forEach(purchaseEntity -> {
            PurchaseDto.Purchase responsePurchaseList = modelMapper.map(purchaseEntity, PurchaseDto.Purchase.class);
            responsePurchaseList.setAccountBookNo(purchaseDto.getAccountBookNo());
            responsePurchaseList.setUserNo(purchaseDto.getUserNo());
            if(purchaseEntity.getCardEntity() != null){
                responsePurchaseList.setCardNo(purchaseEntity.getCardEntity().getCardNo());
            }
            if(purchaseEntity.getCategoryEntity() != null){
                CategoryDto.PurchaseCategoryInfo categoryInfo =  modelMapper.map(purchaseEntity.getCategoryEntity(), CategoryDto.PurchaseCategoryInfo.class);
                if(purchaseEntity.getCategoryEntity().getParentCategory() == null){
                    categoryInfo.setParentCategoryNo(purchaseEntity.getCategoryEntity().getCategoryNo());
                    categoryInfo.setParentCategoryName(purchaseEntity.getCategoryEntity().getCategoryName());
                }else{
                    categoryInfo.setParentCategoryNo(purchaseEntity.getCategoryEntity().getParentCategory().getCategoryNo());
                    categoryInfo.setParentCategoryName(purchaseEntity.getCategoryEntity().getParentCategory().getCategoryName());
                }

                responsePurchaseList.setCategoryInfo(categoryInfo);
            }
            purchaseList.add(responsePurchaseList);
        });
        return purchaseList;
    }

    public Slice<PurchaseDto.Purchase> findPurchaseListOfPage(PurchaseDto purchaseDto, Pageable pageable){

        List<PurchaseEntity> purchaseEntityList = purchaseRepository.findPurchasePageCustom(purchaseDto, pageable);
        List<PurchaseDto.Purchase> purchaseList = new ArrayList<>();
        for (PurchaseEntity purchaseEntity : purchaseEntityList) {
            PurchaseDto.Purchase responsePurchaseList = modelMapper.map(purchaseEntity, PurchaseDto.Purchase.class);
            responsePurchaseList.setAccountBookNo(purchaseDto.getAccountBookNo());
            responsePurchaseList.setUserNo(purchaseDto.getUserNo());
            if (purchaseEntity.getCardEntity() != null) {
                responsePurchaseList.setCardNo(purchaseEntity.getCardEntity().getCardNo());
            }
            if (purchaseEntity.getCategoryEntity() != null) {
                CategoryDto.PurchaseCategoryInfo categoryInfo = modelMapper.map(purchaseEntity.getCategoryEntity(), CategoryDto.PurchaseCategoryInfo.class);
                if (purchaseEntity.getCategoryEntity().getParentCategory() == null) {
                    categoryInfo.setParentCategoryNo(purchaseEntity.getCategoryEntity().getCategoryNo());
                    categoryInfo.setParentCategoryName(purchaseEntity.getCategoryEntity().getCategoryName());
                } else {
                    categoryInfo.setParentCategoryNo(purchaseEntity.getCategoryEntity().getParentCategory().getCategoryNo());
                    categoryInfo.setParentCategoryName(purchaseEntity.getCategoryEntity().getParentCategory().getCategoryName());
                }

                responsePurchaseList.setCategoryInfo(categoryInfo);
            }
            purchaseList.add(responsePurchaseList);
        }

        List<PurchaseDto.Purchase> slicePageResult = getSlicePageResult(purchaseList, pageable.getPageSize());
        return new SliceImpl<>(slicePageResult, pageable, hasPurchaseNext(purchaseList, pageable.getPageSize()));
    }

    @Transactional(readOnly = false, rollbackFor = Exception.class)
    public void deletePurchase(Long userNo, Long purchaseNo) throws Exception{
        PurchaseEntity purchaseEntity = purchaseRepository.findEntityGraphByUserEntity_UserNoAndPurchaseNoAndDeleteYn(userNo, purchaseNo, 'N');
        if(purchaseEntity == null){
            throw new Exception("해당값이 존재하지 않습니다.");
        }
        purchaseEntity.delete();

    }

    public PurchaseDto.ResponsePurchaseDetail findPurchase(Long userNo, Long purchaseNo) {
        return purchaseRepository.findPurchase(userNo, purchaseNo);
    }

    @Transactional(readOnly = false, rollbackFor = Exception.class)
    public void modifyPurchase(PurchaseDto purchaseDto) {
        PurchaseEntity updatePurchaseEntity = purchaseRepository.findEntityGraphByUserEntity_UserNoAndPurchaseNoAndDeleteYn(purchaseDto.getUserNo(), purchaseDto.getPurchaseNo(), 'N');

        updatePurchaseEntity.updatePurchase(purchaseDto);

    }

    private List<PurchaseDto.Purchase> getSlicePageResult(List<PurchaseDto.Purchase> result, int limit) {
        List<PurchaseDto.Purchase> returnValue = new ArrayList<>();
        int cnt = 0;
        for (PurchaseDto.Purchase purchase : result) {
            if(cnt == limit){
                break;
            }
            returnValue.add(purchase);
            cnt++;
        }
        return returnValue;
    }

    private Boolean hasPurchaseNext(List<PurchaseDto.Purchase> result, int limit) {
        return result.size() > limit ? true: false;
    }
}

File: src/main/kotlin/com/hjj/apiserver/service/PurchaseService.kt
package com.hjj.apiserver.service

import com.hjj.apiserver.domain.accountbook.AccountRole
import com.hjj.apiserver.domain.purchase.Purchase
import com.hjj.apiserver.dto.purchase.request.PurchaseAddRequest
import com.hjj.apiserver.dto.purchase.request.PurchaseFindOfPageRequest
import com.hjj.apiserver.dto.purchase.request.PurchaseModifyRequest
import com.hjj.apiserver.dto.purchase.response.PurchaseDetailResponse
import com.hjj.apiserver.dto.purchase.response.PurchaseFindOfPageResponse
import com.hjj.apiserver.repository.accountbook.AccountBookRepository
import com.hjj.apiserver.repository.card.CardRepository
import com.hjj.apiserver.repository.category.CategoryRepository
import com.hjj.apiserver.repository.purchase.PurchaseRepository
import com.hjj.apiserver.repository.user.UserRepository
import com.hjj.apiserver.util.CommonUtils
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Slice
import org.springframework.data.domain.SliceImpl
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class PurchaseService(
    private val accountBookRepository: AccountBookRepository,
    private val categoryRepository: CategoryRepository,
    private val purchaseRepository: PurchaseRepository,
    private val userRepository: UserRepository,
    private val cardRepository: CardRepository,
) {

    @Transactional(readOnly = false, rollbackFor = [Exception::class])
    fun addPurchase(userNo: Long, request: PurchaseAddRequest): Purchase {
        val accountBook = accountBookRepository.findAccountBook(
            userNo = userNo,
            accountBookNo = request.accountBookNo,
        ) ?: throw IllegalArgumentException()

        request.validRequest()


        return purchaseRepository.save(
            Purchase(
                purchaseType = request.purchaseType,
                price = request.price,
                reason = request.reason,
                purchaseDate = request.purchaseDate,
                card = request.cardNo?.let {
                    cardRepository.getById(it) ?: throw IllegalArgumentException()
                },
                category = request.categoryNo?.let {
                    categoryRepository.findCategoryByAccountRole(
                        request.categoryNo,
                        request.accountBookNo,
                        userNo,
                        setOf(AccountRole.OWNER, AccountRole.MEMBER)
                    ) ?: throw IllegalArgumentException()
                },
                user = userRepository.getReferenceById(userNo),
                accountBook = accountBookRepository.getReferenceById(accountBook.accountBookNo),
            )
        )
    }

    fun findPurchasesOfPage(request: PurchaseFindOfPageRequest, pageable: Pageable): Slice<PurchaseFindOfPageResponse> {
        val purchases = purchaseRepository.findPurchasePageCustom(
            request.startDate,
            request.endDate,
            request.accountBookNo,
            pageable
        )

        val purchaseFindOfPageResponseList = purchases.map { purchase ->
            PurchaseFindOfPageResponse(
                purchaseNo = purchase.purchaseNo!!,
                userNo = purchase.user.userNo!!,
                cardNo = purchase.card?.cardNo,
                accountBookNo = purchase.accountBook.accountBookNo!!,
                purchaseType = purchase.purchaseType,
                price = purchase.price,
                reason = purchase.reason,
                purchaseDate = purchase.purchaseDate,
                categoryInfo = purchase.category?.let {
                    PurchaseFindOfPageResponse.PurchaseCategoryInfo(
                        parentCategoryNo = it.parentCategory?.categoryNo,
                        categoryNo = it.categoryNo!!,
                        parentCategoryName = it.parentCategory?.categoryName,
                        categoryName = it.categoryName,
                        categoryDesc = it.categoryDesc,
                        categoryIcon = it.categoryIcon,
                    )
                }
            )
        }.toMutableList()

        val hasNext = purchaseFindOfPageResponseList.size > pageable.pageSize
        return SliceImpl(
            CommonUtils.getSlicePageResult(purchaseFindOfPageResponseList, pageable.pageSize),
            pageable,
            hasNext
        )
    }

    @Transactional(readOnly = false, rollbackFor = [Exception::class])
    fun removePurchase(userNo: Long, purchaseNo: Long) {
        purchaseRepository.findEntityGraphByUser_UserNoAndPurchaseNoAndIsDeleteIsFalse(userNo, purchaseNo)?.delete()
            ?: throw IllegalArgumentException()
    }

    @Transactional(readOnly = false, rollbackFor = [Exception::class])
    fun modifyPurchase(userNo: Long, purchaseNo: Long, request: PurchaseModifyRequest) {
        val purchase = purchaseRepository.findEntityGraphByUser_UserNoAndPurchaseNoAndIsDeleteIsFalse(userNo, purchaseNo)
            ?: throw IllegalArgumentException()

        request.validRequest()

        purchase.updatePurchase(
            request = request,
            card = cardRepository.findByCardNoAndUser_UserNoAndIsDeleteIsFalse(request.cardNo ?: 0, userNo),
            category = categoryRepository.findCategoryByAccountRole(
                categoryNo = request.categoryNo ?: 0,
                accountBookNo = request.accountBookNo,
                userNo = userNo,
                accountRoles = setOf(AccountRole.MEMBER, AccountRole.OWNER)
            ),
        )
    }

    fun findPurchase(userNo: Long, purchaseNo: Long): PurchaseDetailResponse {
        return purchaseRepository.findPurchase(userNo, purchaseNo) ?: throw IllegalArgumentException()
    }

}